### DISCUSSION TITLE
Add support of virtual devices

### DISCUSSION THREAD

**Discussion by hlyi** [TD22057/insteon-mqtt#480] [Ideas]
First, some background information: my setup has a 6-button KPL controlling fanlinc ( buton 3-6 of KPL are radio type and are mapped to fan speed high/med/low/off via insteon scenes). It has been integrated into homeassistant via insteon-mqtt. HA can control the light and fan without any issue. However, the command from HA to Fanlinc doesn't change LEDs on KPL. Based up @kpfleming's feedback. the synchronization hasn't been supported yet.

Here is an idea to workaround this issue: creating virtual devices. Basically, user specifies virtual devices in configuration yaml file, those virtual devices would communicate with MQTT, and the real insteon devices wouldn't communicate MQTT directly. The topology looks like this: MQTT<->Virtual Devices<->Real Devices. In my case, it would be one virtual FanLinc exposed to MQTT, and real KPL and Fanlinc are hidden.  The command from MQTT would translate to scene command sending to both KPL and Fanlinc, And any status change from KPL and Fanlinc is mapped to the virtual Fanlinc status change and send to MQTT.

Some details to be worked out, for example, how to specify the mapping between virtual devices and real devices. I have some idea about it. If this idea was adopted, we can discuss more details.

Thanks for the consideration!

**Comment by krkeegan** [comment]
You can already do this in a couple ways:

1. Using a [Scene](https://github.com/TD22057/insteon-mqtt/blob/master/docs/mqtt.md#scene-triggering) command if supported by the device will alter the state of that device and that device will update its linked devices.
2. Create a [PLM Scene](https://github.com/TD22057/insteon-mqtt/blob/master/docs/scenes.md) that controls both the fan and the KPL.  Then trigger the PLM scene from MQTT.

**Reply by kpfleming** [reply]
Thanks for this... I hadn't spent any time with the scenes support yet, and just set it up on one of my systems. I was able to import all of the existing links, do a bit of trimming/cleanup, and sync back to the devices.

In order to use option 1, when a user activates the light on a FanLinc (using the HA UI), instead of sending directly to group 1 on that FanLinc, we'd want to send a command to group 1 on the associated KeypadLinc instead. That works for the light, but for the fan speed it's trickier because there is only one command endpoint on the FanLinc but there are four on the KeypadLinc.

**Reply by kpfleming** [reply]
So I finally got a chance to try using Scene commands to the various groups on the KPL (which have been linked to the FL using scenes.yaml). Scene commands to group 1 mostly work (the light goes on and off, although a brightness level included in the scene command has no effect), but scene commands to the other group numbers are acknowledged by the device but have no effect on the FL (or even on the KPL button LEDs). I'll need to grab a copy of the Insteon developer's guide to see if this sort of thing is even possible.

**Reply by krkeegan** [reply]
It is hit and miss.  I have a few cases where it matters what firmware version the device has.  Sadly, Insteon was not super consistent with this feature, which is why I say "if supported".

I think your best alternative then is a PLM Scene that controls both the button and the other device.

**Reply by hlyi** [reply]
I ended up creating virtual light and Fan on HA. It is HA specific solution, which is not portable to HE or other  platform. It would be great that insteon-mqtt could add virtual device support, so the solution would be platform agnostic.

**Reply by kpfleming** [reply]
I've added PLM scenes and confirmed that they do light the button on the KPL and also activate the fan, but (unsurprisingly) the radio button grouping on the KPL is not respected. I'm going to set this part of the project aside until I get some time to work on proper radio button support.

**Comment by hlyi** [comment]
Thanks for the suggestion on using Scene. I agree that there could be multiple ways to handle it. One way is what you said, it requires more work in HA. Another way is what I mentioned here, so setup in HA is easier and streamlined via MQTT discovery. IMHO, it would be better that insteon-mqtt hides insteon-specific device handling.  In my use case, it is no appealing reason to expose KPL to HA. A virtual fanlinc as the gateway for both KPL and real fanlinc is a cleaner solution.
